Add Yomi (#109)
uyuni-project/uyuni-docs: modules/salt/pages/yomi.adoc

[[yomi.installer]]
= Install with Yomi

> Yomi (yet one more installer) is an installer for openSUSE operating systems.

Reviewer: Actually works also for SLE

> It is a SaltStack state, and can be used for installing Salt clients.

Reviewer: Yomi is for installing openSUSE, SLE or MicroOS operating systems (similar to autoyast), not for installing Salt.

Yomi consists of two components:

* The Yomi formula, which contains the Salt states and modules required to drive the installation.

> For the formula, see https://build.opensuse.org/package/show/systemsmanagement:yomi/yomi-formula.

Reviewer: This URL is correct. But we managed to add the formula as a package for Uyuni. The user can install it now via `zypper install yomi-formula`

* The LiveCD operating system image, which includes the `pre-configured ``salt-minion`` service.

> For the image, see https://build.opensuse.org/package/show/systemsmanagement:yomi/openSUSE-Tumbleweed-Yomi

Reviewer: This is correct too. Also there is a pxe-yomi-image-opensuse15 package in Uyuni (or will be there soon). I will add installation instructions once the SR is created and accepted.

Author: I'll update when this changes.

To use Yomi for installing your client operating system, you need to follow this process:

> * Install and configure the ``salt-master``service.

Reviewer: In the context of Uyuni this step is already there. In the general case this is needed, tho.

* Install the Yomi formula package.
* Prepare the Salt pillar for the new installation.

> * Boot the new systems with the operating system image, or use PXE boot.

Reviewer: Maybe needs to be pointed that the new machine needs to boot using the Yomi image. Once the pxe-yomi-image gets merged in Uyuni, I will add the instructions.

For more information about using Yomi, see https://github.com/openSUSE/yomi/blob/master/README.md

> == Install and Configure salt-master

Reviewer: In the general case this is required, but because Uyuni already provides a salt-master, this full step is not necessary.

The yomi-formula package takes care also of adding a configuration file snippet to adjust the salt-master configuration.

You will require the ``salt-master`` service to be able to use Yomi for your client operating system.
If you have not yet installed it on the {productname} Server, you will need to do this step first.

.Procedure: Installing salt-master

. On the {productname} Server, at the command prompt, as root, install the ``salt-master`` service:
+
----
zypper in salt-master
----
. Use ``systemctl`` to enable the service:
+
----
systemctl enable --now salt-master.service
----

Yomi uses a Salt pillar to provide configuration information for the installation.
This can include the layout of the hard disks, the software
patterns to install, or the users to create.

By default, Salt searches for states in the [path]``/srv/salt/`` directory.
By default, Salt searches for pillars in [path]``/srv/pillar``.

You do not need to perform this step, if you are using the Yomi formula.

.Procedure: Configuring salt-master

. On the {productname} Server, at the command prompt, as root, create two new directories for the configuration file:
+
----
mkdir -p venv/etc/salt/pki/{master,minion}
----

> .  Create two new virtual environments:

Reviewer: Because this section is part of the salt-master installation, and also not required in Uyuni context, it will not appear at the end. Only to mention that never deploy anything Python-related via a virtual environment : )))

+
----
venv/etc/salt/autosign_grains
venv/var/cache/salt/master/file_lists/roots
----
. Add the configuration information to the new location:
+
----
cat <<EOF > venv/etc/salt/master
root_dir: $(pwd)/venv

file_roots:
  base:
    - $(pwd)/srv/salt

pillar_roots:
  base:
    - $(pwd)/srv/pillar
EOF
----

== Install the Yomi Formula Package

You can use the Yomi formula to install the states and modules you need.

The formula installs the states in [path]``/usr/share/salt-formulas/states``, and configuration files in [path]``/usr/share/yomi``.
It also stores some pillar examples in [path]``/usr/share/yomi/pillar``.

.Procedure: Installing the Yomi Formula

. On the {productname} Server, at the command prompt, as root, install the formula:
+
----
zypper in yomi-formula
----
. So that ``salt-master`` can find the pillar, change the
``pillar_roots`` entry in the configuration file, or use the one provided by the package:
+
----
cp -a /usr/share/yomi/pillar.conf /etc/salt/master.d/

> ----

Reviewer: not need, I hope that the only required step is the installation, the package will take care of the rest in the case of Uyuni.

Edit: The current documentation in the github repository is correct, because is outside of the Uyuni context. Yomi is designed to work alone or integrated. The documentation in github is under the assumption that is working as a single component. This can explain why some steps are not required.

. Restart services:
+
----

> systemctl restart salt-master.service

Reviewer: For safety I will keep the step, tho.

----

> == Prepare the Salt Pillar

Reviewer: This section can be considered the core of the documentation. The user can create the pillar manually, as described here. But because is also using Uyuni, after installing the yomi-formula the user can access to a new set of forms.

Maybe I can prepare some screenshots about how the user can use Uyuni and the forms to create pillars?

Author: Sure, thanks :)

The ``salt-api`` service is required so that Yomi can monitor the installation.

.Procedure: Preparing the Salt Pillar

. Enable this service:
+
----
cp /usr/share/yomi/salt-api.conf /etc/salt/master.d/

> systemctl restart salt-master.service

Reviewer: salt-api is not required for the definition of the pillar, only for monitoring. I will describe the process of how to intall salt-api and use the monitoring tool, but in a separate section about pillars, as are not related.

[[yomi.installer]]
= Install with Yomi

Yomi (yet one more installer) is an installer for openSUSE operating systems.
It is a SaltStack state, and can be used for installing Salt clients.

Yomi consists of two components:

* The Yomi formula, which contains the Salt states and modules required to drive the installation.

> For the formula, see https://build.opensuse.org/package/show/systemsmanagement:yomi/yomi-formula

Reviewer 2: ```suggestion
For the formula, see https://build.opensuse.org/package/show/systemsmanagement:yomi/yomi-formula.
```

* The LiveCD operating system image, which includes the `pre-configured ``salt-minion`` service.
For the image, see https://build.opensuse.org/package/show/systemsmanagement:yomi/openSUSE-Tumbleweed-Yomi

> For the Yomi source code, see https://github.com/openSUSE/yomi.

Reviewer 2: ```suggestion
```
I think this reference is not needed here.  If you think it's important enough, mention it either at the very beginning of this section (right after the intro para) or at the end.

> To use Yomi as your client operating system, you need to follow this process:

Reviewer 2: yomi is not a client OS.

> To use Yomi as your client operating system, you need to follow this process:

Reviewer 2: Maybe this is meant:
```suggestion
To use Yomi for installing your client operating system, you need to follow this process:
```

By default, Salt searches for pillars in [path]``/srv/pillar``.

You do not need to perform this step, if you are using the Yomi formula.

.Procedure: Configuring salt-master

. On the {productname} Server, at the command prompt, as root, create two new directories for the configuration file:
+
----
mkdir -p venv/etc/salt/pki/{master,minion}
----
.  Create two new virtual environments:
+
----
venv/etc/salt/autosign_grains
venv/var/cache/salt/master/file_lists/roots
----
. Add the configuration information to the new location:
+
----
cat <<EOF > venv/etc/salt/master
root_dir: $(pwd)/venv

file_roots:
  base:
    - $(pwd)/srv/salt

pillar_roots:
  base:
    - $(pwd)/srv/pillar
EOF
----

== Install the Yomi Formula Package

You can use the Yomi formula to install the states and modules you need.

The formula installs the states in [path]``/usr/share/salt-formulas/states``, and configuration files in [path]``/usr/share/yomi``.
It also stores some pillar examples in [path]``/usr/share/yomi/pillar``.

.Procedure: Installing the Yomi Formula

. On the {productname} Server, at the command prompt, as root, install the formula:
+
----
zypper in yomi-formula
----

> . So that ``salt-master`` can find the pillar, change the

Reviewer 2: avoid line break here...

``pillar_roots`` entry in the configuration file, or use the one provided by the package:
+
----
cp -a /usr/share/yomi/pillar.conf /etc/salt/master.d/
----
. Restart services:
+
----
systemctl restart salt-master.service
----

== Prepare the Salt Pillar

The ``salt-api`` service is required so that Yomi can monitor the installation.

.Procedure: Preparing the Salt Pillar

. Enable this service:
+
----
cp /usr/share/yomi/salt-api.conf /etc/salt/master.d/
systemctl restart salt-master.service
----
. OPTIONAL: Open the [path]``/etc/salt/master.d/salt-api.conf`` configuration file, and add the required certificates to enable SSL.

> . Open the pillar configuration file (location?) and add or edit these lines:

Reviewer 2: `configuration file (location?)` need further checking.

Author: @aplanas can you help here?

+
----
config:
  events: yes
----

When the pillar is configured, you can launch the ``monitor`` tool from the command prompt.

> The `monitor` tool storez authentication tokens generated by Salt API in a local cache.

Reviewer 2: ```suggestion
The `monitor` tool stores authentication tokens generated by Salt API in a local cache.
```

> This makes future connections to the service simpler, but can sometimes can cause authentication errors.

Reviewer 2: ```suggestion
This makes future connections to the service simpler, but can sometimes cause authentication errors.
```

If you start the ``monitor`` tool with the ``-r`` option, then the local cache is removed before the connection is made, which can avoid authentication problems.

> 

Reviewer 2: ```suggestion
```
avoid single sentence paras

Launch the ``monitor`` tool:
+
----
export SALTAPI_URL=http://localhost:8000
export SALTAPI_EAUTH=file
export SALTAPI_USER=salt
export SALTAPI_PASS=linux

monitor -r -y
----

[[WARNING]]
====
In this example, the password is in plain text.
Do not use this in a production environment.
====

== Boot Clients

Yomi is a set of Salt states that are used to drive the installation of a new operating system.

> To take full control of the system where the installation will be done, you will need to boot from an external system that provides an already configured `salt-minion`, and a set of CLI tools required during the installation.

Reviewer 2: ```suggestion
To take full control of the system where the installation will be done, you will need to boot from an external system that provides an already configured `salt-minion`, and a set of command line tools required during the installation.
```

> You can install YOMI on clients using different mechanisms.

Reviewer 2: ```suggestion
You can install Yomi on clients using different mechanisms.
```
Or use YOMI everywhere (better not, more than 3 uppercase letters are not that user-friendly).

> There is a multi-build image available for download which provides a LiveCD ISO image, and a PXE boot image.

Reviewer 2: ```suggestion
There is a multi-build image available for download, which provides a LiveCD ISO image and a PXE boot image.
```

> You can download the image from https://build.opensuse.org/package/show/systemsmanagement:yomi/openSUSE-Tumbleweed-Yomi

Reviewer 2: ```suggestion
You can download the image from https://build.opensuse.org/package/show/systemsmanagement:yomi/openSUSE-Tumbleweed-Yomi.
```
